storage/store_pool_test.go

```diff
@@ -0,0 +1,272 @@
+		store, ok := sp.stores[storeID]
+		if !ok {
+			t.Fatalf("store %s isn't in the pool's store list", storeID)
+		}
+		exitcode := store.dead
+
+		if exitcode {
+			return nil
+		}
+		return errors.New("store not marked as dead yet")
+	})
+}
+
+// TestStorePoolGossipUpdate ensures that a store is marked as dead after it
+// times out and that it will be revived after a new update is received.
+func TestStorePoolDies(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	stopper, g, sp := createTestStorePool()
+	defer stopper.Stop()
+	sg := newStoreGossiper(g)
+	sg.gossipStores(uniqueStore, t)
+
+	{
+		sp.mu.RLock()
+		store2, ok := sp.stores[2]
+		if !ok {
+			t.Fatalf("store 2 isn't in the pool's store list")
+		}
+		if store2.dead {
+			t.Errorf("store 2 is dead before it times out")
+		}
+		if e, a := 0, store2.timesDied; e != a {
+			t.Errorf("store 2 has been counted dead %d times, expected %d", a, e)
+		}
+		if store2.index == -1 {
+			t.Errorf("store 2 is mot the queue, it should be")
+		}
+		if e, a := 1, sp.queue.Len(); e > a {
+			t.Errorf("wrong number of stores in the queue expected to be at least:%d actual:%d", e, a)
+		}
+		sp.mu.RUnlock()
+	}
+
+	// Timeout store 2.
+	waitUntilDead(t, sp, 2)
+	{
+		sp.mu.RLock()
+		store2, ok := sp.stores[2]
+		if !ok {
+			t.Fatalf("store 2 isn't in the pool's store list")
+		}
+		if e, a := 1, store2.timesDied; e != a {
+			t.Errorf("store 2 has been counted dead %d times, expected %d", a, e)
+		}
+		if store2.index != -1 {
+			t.Errorf("store 2 is in the queue, it shouldn't be")
+		}
+		sp.mu.RUnlock()
+	}
+
+	sg.gossipStores(uniqueStore, t)
+
+	{
+		sp.mu.RLock()
+		store2, ok := sp.stores[2]
+		if !ok {
+			t.Fatalf("store 2 isn't in the pool's store list")
+		}
+		if store2.dead {
+			t.Errorf("store 2 is dead still, it should be alive")
+		}
+		if e, a := 1, store2.timesDied; e != a {
+			t.Errorf("store 2 has been counted dead %d times, expected %d", a, e)
+		}
+		if store2.index == -1 {
+			t.Errorf("store 2 is mot the queue, it should be")
+		}
+		sp.mu.RUnlock()
+	}
+
+	// Timeout store 2 again.
+	waitUntilDead(t, sp, 2)
+	{
+		sp.mu.RLock()
+		store2, ok := sp.stores[2]
+		if !ok {
+			t.Fatalf("store 2 isn't in the pool's store list")
+		}
+		if e, a := 2, store2.timesDied; e != a {
+			t.Errorf("store 2 has been counted dead %d times, expected %d", a, e)
+		}
+		if store2.index != -1 {
+			t.Errorf("store 2 is in the queue, it shouldn't be")
+		}
+		sp.mu.RUnlock()
+	}
+}
+
+// verifyStoreList ensures that the returned list of stores is correct.
+func verifyStoreList(t *testing.T, sp *StorePool, requiredAttrs []string, expected []int) {
+	var actual []int
+	sl := sp.getStoreList(proto.Attributes{Attrs: requiredAttrs}, false)
+	for _, store := range sl.stores {
+		actual = append(actual, int(store.StoreID))
+	}
+	sort.Ints(expected)
+	sort.Ints(actual)
+	if !reflect.DeepEqual(expected, actual) {
+		t.Error(util.ErrorfSkipFrames(1, "expected %+v stores, actual %+v", expected, actual))
```

Comment:
nit: this pattern kind of sucks; why not return an error from here and have the caller call `t.Error(err)`?